frame/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,228 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Stake Tracker Pallet
+//!
+//! The Stake Tracker pallet is used to maintain sorted lists of [`frame_system::Config::AccountId`]
+//! by listening to the events that Staking emits.
+//!
+//! - [`Config`]
+//! - [`Pallet`]
+//!
+//! ## Overview
+//!
+//! The goal of Stake Tracker is to maintain [`SortedListProvider`] sorted list implementations
+//! based on [`SortedListProvider::Score`]. This pallet implements [`OnStakingUpdate`] interface in
+//! order to be able to listen to the events that Staking emits and propagate the changes to said
+//! lists accordingly. It also exposes [`TrackedList`] that adds defensive checks to a subset of
+//! [`SortedListProvider`] methods in order to spot unexpected list updates on the consumer side.
+//! This wrapper should be used to pass the tracked entity to the consumer.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+pub(crate) mod mock;
+#[cfg(test)]
+mod tests;
+
+use frame_election_provider_support::{SortedListProvider, VoteWeight};
+use frame_support::{
+	defensive,
+	traits::{Currency, CurrencyToVote, Defensive},
+};
+pub use pallet::*;
+use sp_staking::{OnStakingUpdate, Stake, StakingInterface};
+
+use sp_std::{boxed::Box, vec::Vec};
+
+/// The balance type of this pallet.
+pub type BalanceOf<T> = <<T as Config>::Staking as StakingInterface>::Balance;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use crate::*;
+	use frame_election_provider_support::{SortedListProvider, VoteWeight};
+	use frame_support::pallet_prelude::*;
+
+	use sp_staking::StakingInterface;
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The same currency type that's used by Staking.
+		type Currency: Currency<Self::AccountId, Balance = BalanceOf<Self>>;
+
+		/// An interface to Staking.
+		type Staking: StakingInterface<AccountId = Self::AccountId>;
+
+		/// A sorted list of nominators and validators, by their stake and self-stake respectively.
+		type VoterList: SortedListProvider<Self::AccountId, Score = VoteWeight>;
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	pub(crate) fn active_stake_of(who: &T::AccountId) -> BalanceOf<T> {
+		T::Staking::stake(&who).map(|s| s.active).unwrap_or_default()
+	}
+
+	pub(crate) fn to_vote(balance: BalanceOf<T>) -> VoteWeight {
+		let total_issuance = T::Currency::total_issuance();
+		<T::Staking as StakingInterface>::CurrencyToVote::to_vote(balance, total_issuance)
+	}
+}
+
+impl<T: Config> OnStakingUpdate<T::AccountId, BalanceOf<T>> for Pallet<T> {
+	fn on_stake_update(who: &T::AccountId, _: Option<Stake<T::AccountId, BalanceOf<T>>>) {
```

Comment:
is `who` controller or stash? We decided to try and move all things in the `StakingApi` and `OnStakingUpdate` to use stash. 

This also makes me wonder: `T::Staking::stake` is queried by the stash, and return the stash to you as well? that does not seem reasonable.

Reply:
Yes, `who` is a stash. Indeed, it does not make a whole lot of sense. `stash` can to be removed from `Stake` struct as it's redundant.